### DISCUSSION TITLE
Dynamic Relationship Join/Origin based on entry

### DISCUSSION THREAD

**Discussion by zikphil** [sqlalchemy/sqlalchemy#9396] [Usage Questions] [answered]
Hello,

So here is my use case: I want to create an Event ORM object and inside that object would include the following 2 columns:
- Related Entity Type
- Related Entity ID

I am wondering how I can create a `relationship()` attribute that would load whatever entity is being referenced in those 2 columns regardless of their table.

**Comment by zzzeek** [comment]
if this is object-> object then this seems like a generic association, all the approaches possible here are at https://docs.sqlalchemy.org/en/20/orm/examples.html#module-examples.generic_associations

**Comment by zikphil** [comment]
Humm isn't that a bit different? Since in this case it would be a dynamic child and not a parent. When I configure the mapper for the `Event` object I can still not indicate what object it will be in the relationship:

```
    class_.addresses = relationship(
        Address,
        primaryjoin=and_(
            class_.id == foreign(remote(Address.parent_id)),
            Address.discriminator == discriminator,
        ),
        backref=backref(
            "parent_%s" % discriminator,
            primaryjoin=remote(class_.id) == foreign(Address.parent_id),
        ),
    )
```

**Reply by zzzeek** [reply]
there's no such thing as "parent" / "child"  , relationships are inherently bidirectional

**Reply by zikphil** [reply]
In the code above, Address is explicitly written but in my use-case I don't know what the relationship object will be until it's loaded. Does that make a difference? Can I just use a generic Model class?

**Reply by zikphil** [reply]
To give you more context of what Im trying to do:

```
class ModelA():
  foo = Column(String...)

class ModelB():
  bar = Column(String...)

class ModelC():
  bar = Column(String...)

class ModelAEvent():
  model_a_id = foreign key

 related_object_type ?= 'ModelB or ModelC or even ModelA'
 related_id ?= 'related object foreign key'

 related_object = relationship that loads whatever object is defined in the related_object_type & related_id columns

```

**Reply by zikphil** [reply]
When I load ModelAEvents, I want to dynamically include the related_object in the query.

**Reply by zzzeek** [reply]
yah if those three classes do not inherit a common superclass, you need to use a separate relationship per class and use `@property` to switch between them.  this is what's in those examples

**Reply by zzzeek** [reply]
relationship can only point to a single table / SELECT (or a UNION etc) or whatever at a time.  cannot change dynamically

**Comment by zikphil** [comment, marked as answer]
Thank you for your help, for the next guy here is how I did it:
```python
class HasRelatedObjectMixin:
    related_type = Column(String, nullable=True)
    related_id = Column(Integer, nullable=True)

    @property
    def related_object(self):
        return getattr(self, "related_%s" % self.related_type)


@event.listens_for(HasRelatedObject, "mapper_configured", propagate=True)
def setup_listener(mapper, class_):
    for table_name, model in get_relatable_objects().items():
        setattr(
            class_,
            f'related_{table_name}',
            relationship(
                model,
                primaryjoin=and_(
                    class_.related_id == foreign(remote(model.id)),
                    class_.related_type == table_name,
                ),
                lazy="selectin",
                viewonly=True,
                uselist=False
            )
        )
```
